### DISCUSSION TITLE
Loading of dynamic relationship

### DISCUSSION THREAD

**Discussion by madsodgaard** [sqlalchemy/sqlalchemy#10164] [Usage Questions]
Hello!

I am using FastAPI and SQLAlchemy together. And I am trying to fetch user-specific information when I load data from the database using SQLAlchemy. So far I have used `query_expression()` and `with_expression()`, but that only works for literal values (as subqueries can only return one column). And I would like to return an entire object, and was wondering if that was possible in some way?

I have a `FitnessPlan` model and when the user wants fetch that model, I would like to return additional information about whether they have a workout that current day. So I have to perform a query to another table `UserWorkout` depending on their user ID, which is why I use `query_expression()`, so I can map it at execution-time. But I ran into the issue that subquries can only return one column. So I tried circumventing that by only returning the ID, and then if I could somehow do a manual `join`, to load the actual record from the other table. But that does not seem possible, because I get the following error:

```
Could not locate any relevant foreign key columns for primary join condition 'workouts.id = NULL' on relationship FitnessPlan.todays_workout.  Ensure that referencing columns are associated with a ForeignKey or ForeignKeyConstraint, or are annotated in the join condition with the foreign() annotation.
``` 

Here is the code:

```python
class FitnessPlan(Base):
    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    todays_workout_id: Mapped[UUID | None] = query_expression()
    todays_workout: Mapped[Workout | None] = relationship(
        primaryjoin="Workout.id == FitnessPlan.todays_workout_id"
    )

def todays_workout_expression(user_id: UUID):
   return (
        select(UserWorkout.workout_id)
        .join(
            models.UserFitnessPlanParticipation,
            UserWorkout.fitness_plan_participation_id
            == models.UserFitnessPlanParticipation.id,
        )
        .where(UserWorkout.user_id == user_id)
        .where(
            models.UserFitnessPlanParticipation.fitness_plan_id == models.FitnessPlan.id
        )
        .where(models.UserFitnessPlanParticipation.is_active == true())
        .where(func.date_trunc("day", UserWorkout.started_at) == datetime.date.today())
    )

# Route code:
def get():
  user_id = "..."

  query = query.options(
                with_expression(
                    FitnessPlan.todays_workout_id,
                    todays_workout_expression(user_id=user_id),
                ),
            )
```

Is this just asking too much of SQLAlchemy?

**Comment by zzzeek** [comment]
hey there -

you don't need a "dynamic" construct here.  Your `todays_workout_expression` function contains the additional WHERE criteria that you would use for a fixed relationship.    You can relate FitnessPlan to a UserWorkout object that is selected using the linkages between all those tables using the general form documented at https://docs.sqlalchemy.org/en/20/orm/join_conditions.html#relationship-to-aliased-class

**Reply by madsodgaard** [reply]
Thanks for your quick answer! I don't think I quite see how I would use that approach in my scenario, how would I define the actual field? Everytime a new user request comes in, it is a different user ID and therefore a different parameter to the expression function to fetch todays workout. That's why I was thinking I need the dynamic construtions here.

```python
exp =  todays_workout_expression(user_id=???)
todays_workout_alias = aliased(Workout, exp)
todays_workout = relationship(todays_workout_alias, primaryjoin=Workout.id==exp.c.id)
```

Does that make sense?

**Reply by zzzeek** [reply]
FitnessPlan is not per-user , is that right?  if you have an application that wishes to display to a User their "fitness plan" and "todays workout", you would want to be selecting from a user-specific entity, which I suppose would be UserFitnessPlanParticipation.

That is, if FitnessPlan id=5 is shared among many users , it makes no sense for there to be a FitnessPlan.todays_workout_for_user_x() SQL entity.     if you wanted a function like that, I'd pass the User object

```py
fp: FitnessPlan
todays_workout_for_user_xyz = fp.todays_workout_for(user_xyz)
```

that method would then call upon user_xyz -> UserFitnessPlanParticipation at the python level.

without the full picture and what actual SQL you are looking to emit and how you are querying data this seems a little vague (why query FitnessPlan and not UserFitnessPlan if you want user-specific objects?)

**Reply by madsodgaard** [reply]
This is right. `UserFitnessPlanParticipation` is what ties `FitnessPlan` and `User` together. What I am trying to do is that, I have a "List fitness plans" endpoint. And for each fitness plan, I want to return todays workout (if they are enrolled). So return a list of fitness plans along with an optional todays workout. I could just do an additional SQL query for each fitness plan retrieved afterwards, as you suggested, but I was trying to see if I could fetch it all in a single query along with the `FitnessPlan`data.

**Reply by zzzeek** [reply]
how come you dont just do `select(FitnessPlan, Workout).where(...) `  ?

There are lots of awkward ways to get the Workout tacked onto the FitnessPlan  (I'd likely use a fake relationship with [additional criteria](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#adding-criteria-to-loader-options) , but there are also event hooks, hybrid attributes, etc) but I just see the overall pattern here as awkward

**Reply by madsodgaard** [reply]
My thought was that I have multiple routes that require the same logic, so in order not to duplicate, it would be nice if it could just be defined as a relationship/field. Also for when I need to convert from the ORM model to the Pydantic model.

How do you define a fake relationship, so you don't get the `Could not locate any relevant foreign key columns for primary join condition 'workouts.id = NULL' on relationship FitnessPlan.todays_workout.  Ensure that referencing columns are associated with a ForeignKey or ForeignKeyConstraint, or are annotated in the join condition with the foreign() annotation.` error?

Again, I might just have to find another way to do this, just wanted to make sure I wasn't missing anything obvious, thanks for the help!

**Reply by zzzeek** [reply]
Just make something up for the relationship, this seems the most simple:

```py
from __future__ import annotations

from sqlalchemy import create_engine, select
from sqlalchemy import ForeignKey
from sqlalchemy.orm import DeclarativeBase, foreign, remote
from sqlalchemy.orm import Mapped, joinedload
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)

    nothing_to_do_with_a: Mapped[NothingToDoWithA] = relationship(
        viewonly=True,
        primaryjoin=lambda: A.id.bool_op("|")(foreign(NothingToDoWithA.id)),
    )


class NothingToDoWithA(Base):
    __tablename__ = "b"
    id: Mapped[int] = mapped_column(primary_key=True)


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

with Session(e) as sess:
    sess.add(A(id=1))
    sess.add(NothingToDoWithA(id=12))

    sess.commit()

    for a1 in sess.scalars(
        select(A).options(
            joinedload(A.nothing_to_do_with_a.and_(NothingToDoWithA.id == 12))
        )
    ):
        print(f"a: {a1} nothing to do with it: {a1.nothing_to_do_with_a}")

```
